lightning/src/util/fairrwlock.rs

```diff
@@ -0,0 +1,42 @@
+use std::sync::{TryLockResult, LockResult, RwLock, RwLockReadGuard, RwLockWriteGuard};
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+/// Rust libstd's RwLock does not provide any fairness guarantees (and, in fact, when used on
+/// Linux with pthreads under the hood, readers trivially and completely starve writers).
+/// Because we often hold read locks while doing message processing in multiple threads which
+/// can use significant CPU time, with write locks being time-sensitive but relatively small in
+/// CPU time, we can end up with starvation completely blocking incoming connections or pings,
+/// especially during initial graph sync.
+///
+/// Thus, we need to block readers when a writer is pending, which we do with a trivial RwLock
+/// wrapper here. Its not particularly optimized, but provides some reasonable fairness by
+/// blocking readers (by taking the write lock) if there are writers pending when we go to take
+/// a read lock.
+pub struct FairRwLock<T> {
```

Comment:
I think fair is somewhat subjective. In this case, it's fair because it prioritizes writing. Should that perhaps be reflected in the name, though I don't really have good suggestions? Like WritePreferenceRwLock? WriterPriorityRwLock? Open to suggestions.

Reply:
I mean its fair in the traditional sense that there is no starvation. Writers won't starve readers, either, I believe, as long as the underlying native RwLock doesn't allow writers to starve other writers.

Reply:
I suppose that's not entirely true, if the underlying RwLock allows writers to starve readers we will still exhibit that behavior here, but I don't think any do, so for now its probably fine, will add a comment.